collections/src/main/scala/strawman/collection/immutable/ChampHashMap.scala

```diff
@@ -0,0 +1,689 @@
+      if (a1(i) != a2(i)) return false
+
+//    // compare local payload
+//    for (i <- Range(0, splitAt)) {
+//      val o1 = a1(i)
+//      val o2 = a2(i)
+//
+//      if (o1 != o2) return false
+//    }
+//
+//    // recursively compare nested nodes
+//    for (i <- Range(splitAt, length)) {
+//      val o1 = a1(i).asInstanceOf[MapNode[K, V]]
+//      val o2 = a2(i).asInstanceOf[MapNode[K, V]]
+//
+//      if (o1 != o2) return false
+//    }
+
+    return true
+  }
+
+}
+
+@SerialVersionUID(2L)
+private[this] final class HashCollisionMapNode[K, +V](val hash: Int, val content: Vector[(K, V)]) extends MapNode[K, V] {
+
+  import MapNode._
+
+  require(content.size >= 2)
+
+  def get(key: K, hash: Int, shift: Int): Option[V] =
+    content.find(key == _._1).map(_._2)
+
+  override def containsKey(key: K, hash: Int, shift: Int): Boolean =
+    content.find(key == _._1).isDefined
+
+  override def contains[V1 >: V](key: K, value: V1, hash: Int, shift: Int): Boolean =
+    content.find(payload => key == payload._1 && value == payload._2).isDefined
+
+  def updated[V1 >: V](key: K, value: V1, hash: Int, shift: Int, effect: MapEffect[K, V1]): MapNode[K, V1] =
+    if (this.contains(key, value, hash, shift)) {
+      this
+    } else if (this.containsKey(key, hash, shift)) {
+      val index = content.indexWhere(key == _._1)
+      val (beforeTuple, fromTuple) = content.splitAt(index)
+      val updatedContent = beforeTuple.appended(Tuple2(key, value)).appendedAll(fromTuple.drop(1))
+
+      effect.setReplacedValue
+      new HashCollisionMapNode[K, V1](hash, updatedContent)
+    } else {
+      effect.setModified
+      new HashCollisionMapNode[K, V1](hash, content.appended(Tuple2(key, value)))
+    }
+
+  def removed[V1 >: V](key: K, hash: Int, shift: Int, effect: MapEffect[K, Any]): MapNode[K, V1] =
+    if (!this.containsKey(key, hash, shift)) {
+      this
+    } else {
+      effect.setModified
+      val updatedContent = content.filterNot(keyValuePair => keyValuePair._1 == key)
+
+      updatedContent.size match {
+        case 0 => MapNode.empty
+        case 1 => new BitmapIndexedMapNode[K, V1]($bitpos($mask(hash, 0)), 0, updatedContent.toArray)
+        case _ => new HashCollisionMapNode[K, V1](hash, updatedContent)
+      }
+    }
+
+  def hasNodes: Boolean = false
+
+  def nodeArity: Int = 0
+
+  def getNode(index: Int): MapNode[K, V] =
+    throw new IndexOutOfBoundsException("No sub-nodes present in hash-collision leaf node.")
+
+  def hasPayload: Boolean = true
+
+  def payloadArity: Int = content.size
+
+  def getPayload(index: Int): (K, V) = content(index)
+
+  def sizePredicate: Int = SizeMoreThanOne
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case node: HashCollisionMapNode[K, V] =>
+        (this eq node) ||
+          (this.hash == node.hash) &&
+            (this.content.size == node.content.size) &&
+            (this.content.filterNot(keyValueTuple => node.content.contains(keyValueTuple))).isEmpty
```

Comment:
This could be optimized even further by using `forall`:

~~~ scala
this.content.forall(node.content.contains)
~~~